pkg/agent/proxy/proxier.go

```diff
@@ -264,6 +275,34 @@ func (p *proxier) installServices() {
 			}
 			endpointUpdateList = append(endpointUpdateList, endpoint)
 		}
+
+		// If the length of endpointUpdateList > maxEndpoints, endpointUpdateList should be cut. However, the iteration
+		// of map in Golang is random, so endpointUpdateList are not always in the same order. If endpointUpdateList is
+		// cut directly without any sorting, some Endpoints may not be installed on the cut endpointUpdateList(Last sync,
+		// these Endpoints may be cut off from endpointUpdateList, they are of course not installed.) So cutting
+		// endpointUpdateList after sorting can avoid this situation in some degree.
+		if len(endpointUpdateList) > maxEndpoints {
+			if !p.oversizeServiceSet.Has(svcPortName.String()) {
+				klog.Warningf("Since Endpoints of Service %s exceeds %d, extra Endpoints will be dropped", svcPortName.String(), maxEndpoints)
+				p.oversizeServiceSet.Insert(svcPortName.String())
+			}
+
+			needUpdateEndpoints = false
+			sort.Sort(byEndpoint(endpointUpdateList))
+			endpointUpdateList = endpointUpdateList[:maxEndpoints]
+			// Check if the cut endpointUpdateList are all installed.
+			for _, endpoint := range endpointUpdateList {
+				if _, ok := endpointsInstalled[endpoint.String()]; !ok {
+					needUpdateEndpoints = true
+					break
+				}
+			}
```

Comment:
Sorting `endpointUpdateList` is to arrange all Endpoints in order and then cut them. If the reserved Endpoints are all installed, `needUpdateEndpoints` will not true. 

More detailed, `needUpdateEndpoints` is recalculated with the cut `endpointUpdateList` and `endpointsInstalled`. `needUpdateEndpoints` will not true if the reserved Endpoints in the cut `endpointUpdateList` are all installed. This part causes extra overhead, but only when the service is oversize.

Reply:
I think L274 will set `needUpdateEndpoints` to true.

Reply:
If a Service is oversize, L274 will always set `needUpdateEndpoints` to true as extra Endpoints are not installed.
L290 resets `needUpdateEndpoints` to false as `endpointUpdateList` is cut and it's unknown that if all reserved Endpoints are installed,  so L294~L299 checks that if there are any Endpoints not install.

Reply:
I missed L290. However, it still seems wrong to reset `needUpdateEndpoints` as there are other conditions that could set it to true, for example, `needUpdateEndpoints = pSvcInfo.SessionAffinityType() != svcInfo.SessionAffinityType()`. And it's redundant to "Check if the cut endpointUpdateList are all installed" twice when it exceeds the size. Could we cut the endpoints in the below place? https://github.com/vmware-tanzu/antrea/blob/eadf0921f88712b373df9f00c778ef81642def2f/pkg/agent/proxy/proxier.go#L244-L246

Reply:
> it still seems wrong to reset needUpdateEndpoints as there are other conditions that could set it to true, for example, `needUpdateEndpoints = pSvcInfo.SessionAffinityType() != svcInfo.SessionAffinityType()`. 

`needUpdateEndpoints = pSvcInfo.SessionAffinityType() != svcInfo.SessionAffinityType()` should be considered.

> Could we cut the endpoints in the below place?

Variable `endpoints` is a map, IMO,  we can't cut the endpoints below the code above.

Reply:
It needs to convert it to a list sooner or later, no harm to do it earlier? It can save the repeated calculation of `needUpdateEndpoints` in your current code L273-L278 and L294-303, and reduce the code complexity in some way I think.

Reply:
Updated.